File: app/src/main/java/com/cristhian/gametobeat/ui/list/GameListActivity.java
package com.cristhian.gametobeat.ui.list;

import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.TabLayout;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;

import com.cristhian.gametobeat.R;
import com.cristhian.gametobeat.database.Game;
import com.cristhian.gametobeat.database.GameViewModel;
import com.cristhian.gametobeat.ui.login.LoginActivity;
import com.cristhian.gametobeat.ui.search.GameSearchActivity;
import com.cristhian.gametobeat.ui.widget.GameWidget;
import com.google.firebase.auth.FirebaseAuth;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class GameListActivity extends AppCompatActivity {

    @BindView(R.id.vp_game_list)
    ViewPager mViewPager;

    @BindView(R.id.tl_game_list)
    TabLayout mTabLayout;

    private FirebaseAuth mAuth;

    private GameViewModel mGameViewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_game_list);

        ButterKnife.bind(this);

        mAuth = FirebaseAuth.getInstance();

        mGameViewModel = ViewModelProviders.of(this).get(GameViewModel.class);

        setupUI();
        setupWidget();
    }

    private void setupUI() {
        GameListPageAdapter pageAdapter = new GameListPageAdapter(getSupportFragmentManager(), this);

        GameListFragment beatGameList = new GameListFragment();
        beatGameList.setTitle(getString(R.string.beat));
        beatGameList.setStatus(Game.STATUS_BEAT);

        GameListFragment playingGameList = new GameListFragment();
        playingGameList.setTitle(getString(R.string.playing));
        playingGameList.setStatus(Game.STATUS_PLAYING);

        GameListFragment wantGameList = new GameListFragment();
        wantGameList.setTitle(getString(R.string.want));
        wantGameList.setStatus(Game.STATUS_WANT);

        pageAdapter.addFragment(beatGameList);
        pageAdapter.addFragment(playingGameList);
        pageAdapter.addFragment(wantGameList);

        mViewPager.setAdapter(pageAdapter);
        mTabLayout.setupWithViewPager(mViewPager);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater menuInflater = new MenuInflater(this);
        menuInflater.inflate(R.menu.menu, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        switch (item.getItemId()) {
            case R.id.logout_action:

                mAuth.signOut();

                Intent intent = new Intent(this, LoginActivity.class);
                startActivity(intent);
                finish();
        }

        return super.onOptionsItemSelected(item);
    }

    @OnClick(R.id.fab_search)
    public void search() {

        Intent intent = new Intent(this, GameSearchActivity.class);
        startActivity(intent);
    }

    private void setupWidget() {
        mGameViewModel.getGamesPlaying().observe(this, new Observer<List<Game>>() {
            @Override
            public void onChanged(@Nullable List<Game> games) {
                GameWidget.mGames = games;
                GameWidget.sendBroadcast(GameListActivity.this);
            }
        });
    }
}

File: app/src/main/java/com/cristhian/gametobeat/ui/list/GameListActivity.kt
package com.cristhian.gametobeat.ui.list

import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import com.cristhian.gametobeat.R
import com.cristhian.gametobeat.database.Game
import com.cristhian.gametobeat.database.GameViewModel
import com.cristhian.gametobeat.ui.login.LoginActivity
import com.cristhian.gametobeat.ui.search.GameSearchActivity
import com.cristhian.gametobeat.ui.widget.GameWidget
import com.google.firebase.auth.FirebaseAuth
import kotlinx.android.synthetic.main.activity_game_list.*

class GameListActivity : AppCompatActivity() {

    lateinit var mGameViewModel: GameViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_game_list)

        mGameViewModel = ViewModelProvider(this).get(GameViewModel::class.java)

        setupUI()
        setupWidget()

        fab_search.setOnClickListener {
            val intent = Intent(this, GameSearchActivity::class.java)
            startActivity(intent)
        }
    }

    private fun setupUI() {
        val pageAdapter = GameListPageAdapter(supportFragmentManager)

        val beatGameList = GameListFragment()
        beatGameList.setTitle(getString(R.string.beat))
        beatGameList.setStatus(Game.STATUS_BEAT)

        val playingGameList = GameListFragment()
        playingGameList.setTitle(getString(R.string.playing))
        playingGameList.setStatus(Game.STATUS_PLAYING)

        val wantGameList = GameListFragment()
        wantGameList.setTitle(getString(R.string.want))
        wantGameList.setStatus(Game.STATUS_WANT)

        pageAdapter.addFragment(beatGameList)
        pageAdapter.addFragment(playingGameList)
        pageAdapter.addFragment(wantGameList)

        vp_game_list.adapter = pageAdapter
        tl_game_list.setupWithViewPager(vp_game_list)
    }

    private fun setupWidget() {
        mGameViewModel.gamesPlaying?.observe(this, Observer {
            GameWidget.mGames = it as List<Game>?
            GameWidget.sendBroadcast(this@GameListActivity)
        })
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        val menuInflater = MenuInflater(this)
        menuInflater.inflate(R.menu.menu, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item!!.itemId) {
            R.id.logout_action -> {
                FirebaseAuth.getInstance().signOut()
                val intent = Intent(this, LoginActivity::class.java)
                startActivity(intent)
                finish()
            }
        }

        return super.onOptionsItemSelected(item)
    }
}
